### ISSUE TITLE
n3fit memory usage

### ISSUE THREAD

**Issue by goord** [NNPDF/nnpdf#1541]
The training of the network seems to consume an awful lot of memory (given the complexity of the task), or at least cluster schedulers seem to perceive the memory use as very high. A NNPDF4.0-NNLO run on the Dutch supercomputer Snellius reports over 200GB of peak memory usage. Detailed measurements will follow.

**Comment by Radonirinaunimi** [comment]
@scarlehoff As far as I recall from previous discussions, the installation was done using the provided environment. However, the tensorflow-gpu did not work out of the box on the A100 (gpus in Snellius) and replaced by a pip installation. This is indeed very large even for a virtual memory allocation. A detailed report (with the fit settings: multi-replicas or not, etc) would surely help understanding what is actually measured.

**Comment by scarlehoff** [comment]
The conda package for tensorflow mkl had a memory leak so that could make the memory grow indefinitely.

That said, if you are running 100 different replicas that could be it. I'm surprised it gets to 200GB though.

Could you share the exact runcard you are using?

**Comment by Radonirinaunimi** [comment]
The run card `NNPDF40_nnlo_as_01180_1000.yml`, from the name, refers to the one in the `reproduce_nnpdf40` folder.

**Comment by scarlehoff** [comment]
Without modifications?

In order to run in GPU for a large number of replicas you need to use some flags as explained here: https://docs.nnpdf.science/n3fit/runcard_detailed.html#running-fits-in-parallel

**Comment by scarlehoff** [comment]
Could you also post the exact version of the code you are running it with and if you have tried with other numbers of replicas?

I don't think I can do 100 without my computer exploding :__ but I might be able to try lower numbers so I can help you troubleshoot better.

**Comment by goord** [comment]
The runcard I've used is [NNPDF40_nnlo_as_01180_1000.yml.txt](https://github.com/NNPDF/nnpdf/files/8260957/NNPDF40_nnlo_as_01180_1000.yml.txt) (I added the .txt extension to keep github happy). I am running the main branch, rev. 0dda92ebb of a few months ago. 

The GPU-equipped nodes I'm running on have 2 sockets with Intel Icelake 8360Y processors (36 cores each) and 4 NVidia A100 40GB cards. The CPU RAM is 512GB (see [here](https://servicedesk.surfsara.nl/wiki/display/WIKI/Snellius+hardware+and+file+systems) for more info)

I have investigated the scaling of the memory usage reported by the SLURM workload manager w.r.t. the no. replicas, both on CPU and on GPU:

| no. replicas | CPU memory reported (GB) | GPU memory reported (GB) | job wall-clock time |
|--------------|--------------------------|--------------------------|---------------------|
| 20           | 45.4                     | 5.2                      | 21 min.                  |
| 40           | 83.3                     | 9.4                      | 32 min.                 |
| 80           | 158.8                    | 18                       | 53 min.                 |

Doing some extrapolation, the compute node should be able to do accommodate about 4x180 fits in parallel, given its GPU configuration, but the RAM usage prevents this. We are currently investigating where all that RAM is being allocated and whether some optimizations can bring that number down a bit.

**Comment by scarlehoff** [comment]
21 minutes for 20 replicas fit is quite good. That's quite some machine you have there!

Now, the biggest thing when doing a fit is not the neural network (that is actually "quite small") but rather the fktables. The idea with the parallel replicas is that the fktables get loaded only once and the same ones used for all fktables (that's why it's important to use the same trvl split) however they get loaded twice, once to generate the covmat and a second time to pass to the fit (why we do that is a long story, but it will be fixed in a future PR when we drop the old `nnfit` machinery).

So my (naive) assumption here is that they are being loaded twice and kept in memory. This would match to first approximation the numbers you see (~2GB per replica).

If that's truly the problem, https://github.com/NNPDF/nnpdf/pull/1529 should fix it. In the next few weeks that PR should become usable (at the moment you would need to install `pineappl` and everything else manually and we still don't have the grids in any available server).

However, *if* that is the problem my expectation is that the memory explosion should be seen already before getting to the fitting stage (so the GPU would not be filled at all at that point) and only grow a bit more after that, is that the case? (in other words, what is the sequence of events here?)

**Comment by goord** [comment]
Hi @scarlehoff it is quite difficult to match memory spikes with events in the code, but to us it indeed seems that the memory increase occurs before the actual fit (perhaps in model generation if I'm correct). Let me check if I can produce a timeline...

**Comment by scarlehoff** [comment]
Ok, don't worry. Having just a rough idea will be great (basically we want to know whether it happens due to the fit or is the loading of the data and theory that is the problem).

> but to us it indeed seems that the memory increase occurs before the actual fit

This is a "best case scenario" because it's something that might get fixed "organically" once we start using `pineappl` :)

**Comment by Radonirinaunimi** [comment]
@scarlehoff Do you have a sort of timeline on when #1529 will be merged? But if in the meantime you, @goord, want to check #1529, I can provide you with some `pineappl` grids.

**Comment by scarlehoff** [comment]
By the end of the month during the code meeting.

But by the end of this week I want to have it in a fully functional state (we are fixing the last problems we have with some of the grids / operations etc) and once it is fully functional I will also upload the grids we are using to the nnpdf server so they should be available.

But what I want and what I will be able to do are not necessarily the same thing :P

**Comment by scarlehoff** [comment]
One thing that just came to mind, the memory that is being reported is the actual RAM used?

Because the `VIRT` size (when looking with `htop`) is usually even an order of magnitude above the actual RAM used. Maybe `slurm` is just being too strict?

I'm trying in my computer with 1-4 replicas and the actual ram memory used during the fit doesn't grow much but the `VIRT` size starts growing beyond what my computer can do.

**Comment by goord** [comment]
Hi @scarlehoff so that's one thing I'm trying to figure out, and I mentioned above that the slurm manager may be wrong on the actual RAM usage. I am now looking at a fit of 100 replicas on the CPU. It uses ~214 GB of virtual memory, ~195 GB of resident memory using htop. The libmemusage tool reports 195GB peak memory usage, so for this problem size the virtual memory overhead seems to be reasonably small

**Comment by scarlehoff** [comment]
Ok, so it's not that then. It would be great to understand whether it is happening before or after the fit then.

If you stop the code shortly after it gets inside `performfit` (in `performfit.py`) the memory usage you see should be only due to the loading of the fktables, masks and data.

**Comment by scarlehoff** [comment]
Actually, sorry I was wrong before. @goord, @Radonirinaunimi  you should be able to use https://github.com/NNPDF/nnpdf/pull/1529 already since when you don't have `pineappl` installed it will fallback to using the old fktables but with the renewed code so you should see an improvement already on that one.

**Comment by scarlehoff** [comment]
Hi @goord many thanks for your report! I think I've found and isolated the problem. Hopefully will be able to fix it during next week (in #1529, in the current master it would be harder)

**Comment by scarlehoff** [comment]
The problem should be fixed in #1529 (and given that the fit bot seem to have worked ok, everything should be fine).

**Comment by Radonirinaunimi** [comment]
> The problem should be fixed in #1529 (and given that the fit bot seem to have worked ok, everything should be fine).

Thanks @scarlehoff, we will have a look.

**Comment by goord** [comment]
@scarlehoff great, I will test with the branch this evening

**Comment by scarlehoff** [comment]
> @scarlehoff great, I will test with the branch this evening

did you have any change to try this? (I would like to merge that branch asap so if you had already some feedback it would be very helpful!)

**Comment by goord** [comment]
Hi @scarlehoff sorry for the delay, I was on a conference and now the supercomputer is down for maintenance until tomorrow evening. I can report my findings after the weekend

**Comment by goord** [comment]
Hi @scarlehoff and @Radonirinaunimi the ``pinappl`` branch indeed shows a big reduction in memory footprint, initial (cpu-only) tests 

| no. replicas | peak heap (GB) | job wall-clock time |
|--------------|--------------------------|--------------------------|
| 20            | 10.8                       | 55 min.                  |
| 40            | 15.0                       | 101 min.                |
| 60            | 19.5                       | 165 min.                |
| 80            | 23.6                       | 223 min.                |

GPU tests are underway.

**Comment by Radonirinaunimi** [comment]
> Hi @scarlehoff and @Radonirinaunimi the `pinappl` branch indeed shows a big reduction in memory footprint, initial (cpu-only) tests
> no. replicas 	peak heap (GB) 	job wall-clock time
> 20 	10.8 	55 min.
> 40 	15.0 	101 min.
> 60 	19.5 	165 min.
> 80 	23.6 	223 min.
> 
> GPU tests are underway.

These are incredible improvements. Thanks for these benchmarks, looking forward to the GPU results.

**Comment by scarlehoff** [comment]
Wow, even better than I expected :) Thank you very much for the benchmarks!

**Comment by goord** [comment]
First results. Runtime is slightly slower, but memory is again drastically reduced:

| no. replicas | Peak heap RAM (GB) | GPU memory reported (GB) | job wall-clock time |
|--------------|--------------------------|--------------------------|---------------------|
| 20            | 12                          | 4.8                         | 25 min.            |
| 40            | 16                          | 8.8                         | 44 min.            |
| 80            | 25                          | 17                          | 60 min.            |

Now GPU memory is the limiting factor (as usual). The machine I'm running on has actually 4 cards per node, so I think it would be useful to investigate distributed training no?

**Comment by scarlehoff** [comment]
How does it scale with the number of epochs? 1000 epochs is actually not much, a fit will have order 20k, I'm guessing at the beginning it won't be linear (since there's some time spent on loading the data and generating replicas).

After that, it would be more interesting to improve the performance in a single-GPU since having 4 GPUs in a single computer is not the usual scenario.

**Comment by goord** [comment]
Absolutely, large part of the job time is initialization, I can insert timers for the actual fit. In terms of memory, I don't observe a significant increase during the fit itself

**Comment by Zaharid** [comment]
FWIW it would be nice to see how things look using this thing

https://github.com/bloomberg/memray

**Comment by Radonirinaunimi** [comment]
> FWIW it would be nice to see how things look using this thing
> 
> https://github.com/bloomberg/memray

Indeed! I saw that this was made open-source two weeks ago but forgot to mention it. This would be one of rarest good and reasonable python memory profiling that exist today.

Also, @goord, as we discussed, it would be very useful to have the same benchmarks as above but with a full fit.
